src/os_cloud_nodelet.cpp

```diff
@@ -33,6 +34,30 @@ using ouster_ros::PacketMsg;
 using sensor::UDPProfileLidar;
 using namespace std::chrono_literals;
 
+namespace {
+template <typename T, typename UnaryPredicate>
+int find_if_reverse(const Eigen::Array<T, -1, 1>& array,
+                    UnaryPredicate predicate) {
+    auto p = array.data() + array.size() - 1;
+    do {
+        if (predicate(*p)) return p - array.data();
+    } while (p-- != array.data());
+    return -1;
+}
+
+template <typename X, typename Y>
+Y linear_interpolate(X x0, Y y0, X x1, Y y1, X x) {
+    return y0 + (x - x0) * (y1 - y0) / (x1 - x0);
+}
+
+template <typename T>
```

Comment:
`T` is only `double` at the moment - does this need to be a template? (If so, is there an easy check to make sure that `static_cast<T>(ULLONG_MAX)` would not overflow?)

Reply:
It does not need to be a template, we can make as `double` only method, i just thought it would be appropriate to implement it as template since `std::round` accepts `float` numbers too. I don't know if captured the second part of the question correctly but the idea for me was to keep `ulround` generic that it can handle an non-integral singed data value and convert it properly/safely into `uint64_t` regardless of the range of values you may pass to it.

Reply:
My guess is that the method isn't meant to handle signed integrals, but for example `ulround(static_cast<short>(16000))` yields `0`.

Reply:
You are right it, I didn't actually try the method against integer types.

Reply:
I think? that we only need it for integer types? because a roundtrip through double and rounding seems unideal to me

Reply:
Addressed the issue with function not working for integer data types.

@kairenw why is it not ideal? can you elaborate?

Reply:
UTC in nanoseconds is large and I think we'd lose like 2-3 decimal points precision

Since I don't see that we'd ever be interpolating between non-timestamp values which will always be `uint64_t`, it makes more sense to me to take `uint64_t` ts arguments, subtract out the min of those for the interpolation, and add it back on before returning a `uint_64t`

Reply:
Makes sense. Addressed